test/compile-error-test/CMakeLists.txt

```diff
@@ -1,77 +1,128 @@
 # Test if compile errors are produced where necessary.
 
 cmake_minimum_required(VERSION 3.1...3.18)
+project(compile-error-test)
 
-include(CheckCXXSourceCompiles)
-include(CheckCXXCompilerFlag)
+set(test_names "")
 
-set(CMAKE_REQUIRED_INCLUDES ${CMAKE_CURRENT_SOURCE_DIR}/../../include)
-set(CMAKE_REQUIRED_FLAGS ${CXX_STANDARD_FLAG} ${PEDANTIC_COMPILE_FLAGS})
-
-function (generate_source result fragment)
-  set(${result} "
-  #define FMT_HEADER_ONLY 1
-  #include \"fmt/format.h\"
-  int main() {
-    ${fragment}
-  }
-  " PARENT_SCOPE)
+function (expect_compile name code_fragment)
+  cmake_parse_arguments(EXPECT_COMPILE "ERROR;FORMAT;XCHAR" "" "" ${ARGN})
+  set(test_names_local_copy "${test_names}")
+  list(APPEND test_names_local_copy "${name}")
+  set(test_names "${test_names_local_copy}" PARENT_SCOPE)
+  set(header "fmt/core.h")
+  if (EXPECT_COMPILE_FORMAT)
+    set(header "fmt/format.h")
+  elseif(EXPECT_COMPILE_XCHAR)
+    set(header "fmt/xchar.h")
+  endif()
+  configure_file("${CMAKE_CURRENT_SOURCE_DIR}/source.cpp.in" "${CMAKE_CURRENT_BINARY_DIR}/tests/${name}.cpp")
```

Comment:
I suggest using `file(CONFIGURE ...)` instead since the template is trivial and having it inline makes the logic easier to understand.

Reply:
``file(CONFIGURE ...)`` - present only begin from CMake 3.18 (https://cmake.org/cmake/help/v3.18/release/3.18.html#commands).

Reply:
I guess we are leaving it as it is then, right? 😉